Sources/SwiftProtobuf/JSONScanner.swift

```diff
@@ -14,245 +14,335 @@
 ///
 // -----------------------------------------------------------------------------
 
-private func fromHexDigit(_ c: Character?) -> UInt32? {
-    if let c = c {
-        switch c {
-        case "0": return 0
-        case "1": return 1
-        case "2": return 2
-        case "3": return 3
-        case "4": return 4
-        case "5": return 5
-        case "6": return 6
-        case "7": return 7
-        case "8": return 8
-        case "9": return 9
-        case "a", "A": return 10
-        case "b", "B": return 11
-        case "c", "C": return 12
-        case "d", "D": return 13
-        case "e", "E": return 14
-        case "f", "F": return 15
-        default: return nil
-        }
+public class JSONScanner {
+
+  /// A stack of tokens that should be returned upon the next call to `next()`
+  /// before the input string is scanned again.
+  private var tokenPushback: [JSONToken]
+
+  /// The scanner uses the UnicodeScalarView of the string, which is
+  /// significantly faster than operating on Characters.
+  private let scalars: String.UnicodeScalarView
```

Comment:
I'm curious:  Is there additional memory overhead for this?   In particular, I'm not sure if `String.unicodeScalars` has to copy the original.